mavisp/modules.py

```diff
@@ -768,6 +767,109 @@ def ingest(self, mutations):
+            raise MAVISpMultipleError(warning=warnings,
+                                    critical=[MAVISpCriticalError(this_error)])
+
+        try:
+            # Data type alignment and merge
+            sas_data = sas_data.reset_index()
+            aggregated_df = pd.merge(aggregated_df, sas_data, left_on='resnum', right_on='residue', how='left')
+            aggregated_df['restype_resnum_kinase'] = aggregated_df['restype'] + aggregated_df['resnum'].astype(str) + '_' + aggregated_df['kinase']
+        except Exception as e:
+            this_error = f"Error during data preparation: {e}"
+            raise MAVISpMultipleError(warning=warnings,
+                                    critical=[MAVISpCriticalError(this_error)])
+
+        gain_of_function = defaultdict(list)
+        loss_of_function = defaultdict(list)
+
+        # Main analysis loop
+        try:
+            for index, row in aggregated_df.iterrows():
+                for mutation in mutations:
+                    restype_resnum_kinase = row['restype_resnum_kinase']
+                    if pd.isna(row['WT']) and not pd.isna(row[mutation]) and row['sas_sc_rel'] > 20:
+                        gain_of_function[mutation].append(restype_resnum_kinase)
+                    elif not pd.isna(row['WT']) and pd.isna(row[mutation]):
+                        loss_of_function[mutation].append(restype_resnum_kinase)
+        except Exception as e:
+            this_error = f"Error during mutation analysis: {e}"
+            raise MAVISpMultipleError(warning=warnings,
+                            critical=[MAVISpCriticalError(this_error)])
+
+        # Convert the dictionaries into DataFrames
+        try:
+            gain_of_function_df = pd.DataFrame(
+                [(k, ','.join(v)) for k, v in gain_of_function.items()],
+                columns=['mutation', 'PTM gain of function']).set_index('mutation')
+            loss_of_function_df = pd.DataFrame(
+                [(k, ','.join(v)) for k, v in loss_of_function.items()],
+                columns=['mutation', 'PTM loss of function']).set_index('mutation')
+            final_table = pd.DataFrame({'mutation': mutations}).set_index('mutation')
+            final_table = final_table.join(loss_of_function_df, how='left').join(gain_of_function_df, how='left')
+            # Add 'PTM effect by mutation' based on gain or loss of function
+            final_table['Mutation predicted to add new phosphorylation site'] = final_table['PTM gain of function'].apply(lambda x: False if pd.isna(x) else True)
+            self.data = final_table
+        except Exception as e:
+            this_error = f"Error compiling results: {e}"
+            raise MAVISpMultipleError(warning=warnings,
+                                    critical=[MAVISpCriticalError(this_error)])
+
+        if len(warnings) > 0:
+            raise MAVISpMultipleError(warning=warnings, critical=[])
```

Comment:
overall, this part looks like a lot like the simple mode `ingest` function. Do you think it would be possible to just not overload the `ingest` function? If your `self._parse_sas` returns a similar dataframe to the simple mode function, you could just reuse it as-is